shutter/low-level.md

```diff
@@ -447,54 +448,63 @@ def get_participating_validators(state) -> Sequence[ValidatorIndex]:
                 version,
                 chain_id,
                 validator_registry_address,
-                validator_index,
+                start_index,
+                count,
                 nonce,
                 is_registration,
             ) = extract_message_parts(update.message)
         except:
             continue
 
+        if count == 0:
+            continue
         if version != VALIDATOR_REGISTRY_MESSAGE_VERSION:
             continue
         if chain_id != CHAIN_ID:
             continue
         if validator_registry_address != VALIDATOR_REGISTRY_ADDRESS:
             continue
 
-        if validator_index >= len(state.validators):
-            continue
-        if nonce <= prev_nonces.get(validator_index, -1):
+        end_index = start_index + count
+
+        if end_index > len(state.validators):
             continue
 
-        validator_pubkey = state.validators[validator_index].pubkey
-        if not bls.Verify(validator_pubkey, keccak256(message), signature):
+        validator_pubkeys = [v.pubkey for v in state.validators[start_index : end_index]]
+        if not bls.VerifyAggregate(validator_pubkeys, keccak256(message), signature):
             continue
 
-        prev_nonces[validator_index] = nonce
-        if is_registration:
-            indices.add(validator_index)
-        else:
-            indices.remove(validator_index)
+        for validator_index in range(start_index, end_index):
+            if nonce <= prev_nonces.get(validator_index, -1):
+                continue
+
+            prev_nonces[validator_index] = nonce
+            if is_registration:
+                indices.add(validator_index)
+            else:
+                indices.remove(validator_index)
 
     return sorted(indices)
 
-def extract_message_parts(message: bytes) -> tuple[bytes, uint64, Address, uint64, uint64, bool]:
+def extract_message_parts(message: bytes) -> tuple[bytes, uint64, Address, uint64, uint32, uint32, bool]:
     suffix = message[-1:]
     is_registration = bool(int.from_bytes(suffix, "big"))
 
     prefix = message[:-1]
-    assert len(prefix) == 1 + 8 + 20 + 8 + 8
+    assert len(prefix) == 1 + 8 + 20 + 8 + 4 + 4
     version = prefix[0:1]
     chain_id_bytes = prefix[1:9]
     registry_address = prefix[9:29]
     index_bytes = prefix[29:37]
-    nonce_bytes = prefix[37:45]
+    count_bytes = prefix[37:41]
+    nonce_bytes = prefix[41:45]
```

Comment:
We're planning to stay backward compatible by allowing legacy `0x00` and new `0x01` messages within the same deployment. Here is the only point where I can see trouble: if any validator used a `nonce > MAX_UINT_32` in the past, they would not be able to switch to the new format.

Reply:
As discussed offline we'll not keep backwards compatibility to keep the spec simple, since we don't have many existing validators

Reply:
With new change we slightly alter V0 assuming the nonce only takes 4 bytes, the upper 4 bytes are ignored